### PR TITLE
Build ingress-nginx for WAF and arm64

### PR DESCRIPTION
**Pull request by cataphract**
Related Jira: [APPSEC-55744]

[APPSEC-55744]: https://datadoghq.atlassian.net/browse/APPSEC-55744?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/security/ddwaf_obj.h
[the remark this answers is not in the document]
Comment (cataphract) [reply src/security/ddwaf_obj.h:1]: not at all, it's a compilation warning I noticed. I put it in a separate commit

File: example/ingress-nginx/k8s/configmap.yaml [hunk 1]
```diff
@@ -1,8 +1,16 @@
 kind: ConfigMap
 apiVersion: v1
 metadata:
-  name: nginx-ingress-controller
-  namespace: default
+  name: ingress-nginx-controller
+  namespace: ingress-nginx
```
[the remark this answers is not in the document]
Comment (cataphract) [reply example/ingress-nginx/k8s/configmap.yaml:5]: these are not unnecessary. The ingress-nginx configuration [mentioned in the README](https://raw.githubusercontent.com/kubernetes/ingress-nginx/main/deploy/static/provider/cloud/deploy.yaml) uses this name and namespace:

```yaml
---
apiVersion: apps/v1
kind: Deployment
metadata:
  ...
  name: ingress-nginx-controller
  namespace: ingress-nginx
spec:
  ...
  template:
    ...
    spec:
      containers:
      - args:
        - /nginx-ingress-controller
        - --publish-service=$(POD_NAMESPACE)/ingress-nginx-controller
        ...
```

basically we're aiming to override the `ConfigMap` included in that deployment file:

```yaml
---
apiVersion: v1
data: null
kind: ConfigMap
metadata:
  labels:
    app.kubernetes.io/component: controller
    app.kubernetes.io/instance: ingress-nginx
    app.kubernetes.io/name: ingress-nginx
    app.kubernetes.io/part-of: ingress-nginx
    app.kubernetes.io/version: 1.12.0-beta.0
  name: ingress-nginx-controller
  namespace: ingress-nginx
```

Also, btw, [_nginx-ingress_](https://github.com/nginxinc/kubernetes-ingress) is a different product from [_ingress nginx_](https://github.com/kubernetes/ingress-nginx)

File: example/ingress-nginx/README.md [hunk 2]
```diff
@@ -33,10 +33,39 @@ spec:
     spec:
       initContainers:
         - name: init-datadog
-          image: datadog/ingress-nginx-injection:v1.10.0
+          # choose tag as appropriate, e.g. v1.11.3-appsec-amd64, v1.11.3-arm64
+          image: datadog/ingress-nginx-injection:v1.10.0-amd64
```
[the remark this answers is not in the document]
Comment (cataphract) [reply example/ingress-nginx/README.md:37]: One more reason to keep it like this then. The way it's usually done is you push an -amd64 and an -arm64 tag, and then you push descriptor referencing these two. I can add it to `release.py`

File: .circleci/src/@jobs.yml [hunk 3]
```diff
@@ -173,6 +173,11 @@ jobs:
         type: string
       arch:
         type: string
+      waf:
```
[the remark this answers is not in the document]
Comment (cataphract) [reply .circleci/src/@jobs.yml:176]: sounds good

File: example/ingress-nginx/k8s/configmap.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (dmehala) [reply example/ingress-nginx/k8s/configmap.yaml:5]: My bad, I forgot it was the example. Good catch.

File: example/ingress-nginx/README.md
Comment (dmehala) [inline example/ingress-nginx/README.md:1]: Not related to your changes. ingress-nginx maintainers moved the yaml file. While you are here, may I ask to update the deploy command line to `https://raw.githubusercontent.com/kubernetes/ingress-nginx/refs/tags/controller-v1.11.3/deploy/static/provider/cloud/deploy.yaml`?

File: example/ingress-nginx/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (dmehala) [reply example/ingress-nginx/README.md:37]: Not sure to get it but I think we are saying the same thing?
We should push -amd and -arm64 and the descriptor will dispatch to the right arch image. [Here](https://hub.docker.com/repository/docker/datadog/docker-library/tags?name=httpd-datadog-ci-2.4-e5dd7ac) is an example.

Screenshot 2024-11-15 at 13 23 47
com/repository/docker/datadog/docker-library/tags?name=httpd-datadog-ci-2.4-e5dd7ac
 
Referencing `httpd-datadog-ci-2.4-e5dd7ac` is enough, docker will pull the right os/arch depending of the host arch.

**Review by dmehala — APPROVED** [review]
Noice. I don't know if you are aware that ingress-nginx package a WAF module called [ModSecurity](https://github.com/owasp-modsecurity/ModSecurity) using the [nginx connector](https://github.com/owasp-modsecurity/ModSecurity-nginx?tab=readme-ov-file). It worth investigating the interactions between our module and the other one or/and add instructions to disable modsecurity if they can't be used in parallel.

The extra mile is probably to support _if possible_ drop in support for modsecurity directives.
:shipit:

File: example/ingress-nginx/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (cataphract) [reply example/ingress-nginx/README.md:37]: You're right, for the example it should not have the arch suffix

**Review by dmehala — APPROVED** [review]
LGTM :shipit:

**Pull request merged**